lib/DefinePlugin.js

```diff
@@ -98,25 +120,28 @@ class DefinePlugin {
 									 * });
 									 */
 									if (recurse) return;
+									registerFileDependencies(code);
 									recurse = true;
-									const res = parser.evaluate(code);
+									const res = parser.evaluate(toCode(code));
 									recurse = false;
 									res.setRange(expr.range);
 									return res;
 								});
-							parser.hooks.expression
-								.for(key)
-								.tap(
-									"DefinePlugin",
-									/__webpack_require__/.test(code)
-										? ParserHelpers.toConstantDependencyWithWebpackRequire(
-												parser,
-												code
-											)
-										: ParserHelpers.toConstantDependency(parser, code)
-								);
+							parser.hooks.expression.for(key).tap("DefinePlugin", expr => {
+									registerFileDependencies(code);
+									code = toCode(code);
+
+									if (/__webpack_require__/.test(code)) {
+										return ParserHelpers.toConstantDependencyWithWebpackRequire(
+											parser,
+											code
+										);
+									} else {
+										return ParserHelpers.toConstantDependency(parser, code);
+									}
+								});
 						}
-						const typeofCode = isTypeof ? code : "typeof (" + code + ")";
+						const typeofCode = isTypeof ? toCode(code) : "typeof (" + toCode(code) + ")";
```

Comment:
This need to be moved into the `evaluateTypeof` hook, to be evaluated when requested.